File: binary-tree-bfs/zig-zag-traversal/ZigZagTraversal.java
// https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/?envType=study-plan-v2&envId=top-interview-150

import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) return result;

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        boolean leftToRight = true; // direction flag

        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            Deque<Integer> level = new LinkedList<>();

            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();

                // Add to front or back depending on direction
                if (leftToRight) {
                    level.addLast(node.val);
                } else {
                    level.addFirst(node.val);
                }

                if (node.left != null) queue.offer(node.left);
                if (node.right != null) queue.offer(node.right);
            }

            result.add(new ArrayList<>(level));
            leftToRight = !leftToRight; // flip direction for next level
        }

        return result;
    }
}

File: binary-tree-bfs/zig-zag-traversal/ZigZagTraversal.py
# https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/?envType=study-plan-v2&envId=top-interview-150

from typing import List, Optional
from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
        
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:

        if not root:
            return []
        
        node_queue = deque([root])
        output = []
        reverse = False

        while node_queue:
            node_queue_size = len(node_queue)
            temp_level_arr = []

            for i in range(node_queue_size):
                if reverse: 
                    node = node_queue.pop()
                else:
                    node = node_queue.popleft()

                if reverse: 
                    if node.right: 
                            node_queue.appendleft(node.right)
                    if node.left:
                            node_queue.appendleft(node.left)
                else:
                    if node.left:
                            node_queue.append(node.left)
                    if node.right:  
                            node_queue.append(node.right)

                temp_level_arr.append(node.val)

            output.append(temp_level_arr)
            reverse = not reverse

        return output